src/tests/tracing/eventpipe/applystartuphook/ApplyStartupHookValidation.cs

```diff
@@ -0,0 +1,111 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.Diagnostics.Tools.RuntimeClient;
+using Tracing.Tests.Common;
+using System.Text;
+using System.Threading;
+using System.IO;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+
+namespace Tracing.Tests.ApplyStartupHookValidation
+{
+    public class ApplyStartupHookValidation
+    {
+        public static async Task<bool> TEST_ApplyStartupHookAtStartupSuspension()
+        {
+            bool fSuccess = true;
+            string serverName = ReverseServer.MakeServerAddress();
+            Logger.logger.Log($"Server name is '{serverName}'");
+            Task<bool> subprocessTask = Utils.RunSubprocess(
+                currentAssembly: Assembly.GetExecutingAssembly(),
+                environment: new Dictionary<string,string> 
+                {
+                    { Utils.DiagnosticPortsEnvKey, serverName }
+                },
+                duringExecution: async (_) =>
+                {
+                    ReverseServer server = new ReverseServer(serverName);
+                    using (Stream stream = await server.AcceptAsync())
+                    {
+                        IpcAdvertise advertise = IpcAdvertise.Parse(stream);
+                        Console.WriteLine($"IpcAdvertise: {advertise}");
+
+                        int processId = (int)advertise.ProcessId;
+
+                        // While we are paused in startup, apply startup hook
+                        DiagnosticsClient client = new DiagnosticsClient(processId);
+
+                        string startupHookPath = Hook.Basic.AssemblyPath;
+                        Console.WriteLine($"Applying startup hook during startup suspension: {startupHookPath}");
+                        client.ApplyStartupHook(startupHookPath);
```

Comment:
This is throwing an exception when running on Mono because it's not supported:
```
Microsoft.Diagnostics.NETCore.Client.ServerErrorException: ApplyStartupHook - Not supported by this runtime.
         at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ValidateResponseMessage(IpcMessage responseMessage, String operationName, ValidateResponseOptions options)
         at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.ApplyStartupHook(String startupHookPath)
         at Tracing.Tests.ApplyStartupHookValidation.ApplyStartupHookValidation.<>c__DisplayClass0_0.<<TEST_ApplyStartupHookAtStartupSuspension>b__0>d.MoveNext()
```

Is there a way to detect whether this is running in Mono vs CoreClr (then I can test that it does throw the expected exception on Mono) or possibly to only run this test for CoreClr?

Reply:
For detection at run-time, you can add a reference to `CoreCLRTestLibrary` and use `TestLibrary.Utilities.IsMonoRuntime`:
https://github.com/dotnet/runtime/blob/fc486b4087bd3a6c6c2934da8b3c22cb4bf73523/src/tests/Common/CoreCLRTestLibrary/Utilities.cs#L95

To disable a test entirely, you can update the project to be disabled based on the runtime. For example:
https://github.com/dotnet/runtime/blob/fc486b4087bd3a6c6c2934da8b3c22cb4bf73523/src/tests/tracing/runtimeeventsource/runtimeeventsource.csproj#L10

Reply:
I'd also mark the test nativeaot-incompatible:
```xml
<!-- Startup hooks are not supported -->
<NativeAotIncompatible>true</NativeAotIncompatible>
```

Reply:
I ended up excluding running this test on Mono.